REGRESSION(286883@main): ASSERT NOT REACHED in WebCore::toResourceLoadPriority(CFURLRequestPriority)

pedrovarangot opened the pull request:
#### 3a885897a634abba8d8c80b086bd8eb0fe1ede39
<pre>
REGRESSION(286883@main): ASSERT NOT REACHED in WebCore::toResourceLoadPriority(CFURLRequestPriority)
<a href="https://rdar.apple.com/140442183">rdar://140442183</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=283596">https://bugs.webkit.org/show_bug.cgi?id=283596</a>

Reviewed by Ryosuke Niwa.

This fixes the regression by setting a value for the priority that is within bounds, and prevents future asserts by clamping

* LayoutTests/ipc/cfnetwork-crashes-with-string-to-string-http-headers.html:
* LayoutTests/ipc/url-requestPriority-out-of-range-expected.txt: Added.
* LayoutTests/ipc/url-requestPriority-out-of-range.html: Copied from LayoutTests/ipc/cfnetwork-crashes-with-string-to-string-http-headers.html.
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
(WebKit::CoreIPCNSURLRequest::toID const):

Canonical link: <a href="https://commits.webkit.org/289501@main">https://commits.webkit.org/289501@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41b3645fcef327cb9e31970737bf5d4e7a667974

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6636 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41485 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91985 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37865 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6912 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14704 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67336 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25089 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90128 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5287 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78859 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47655 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5060 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33236 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36982 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75549 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34114 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93872 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14288 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76138 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14492 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74715 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75339 "Found 1 new API test failure: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18547 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19679 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18116 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/7205 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14307 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19600 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14052 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17495 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15833 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -250,7 +250,9 @@ - (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
     [dict setObject:[NSNumber numberWithUnsignedChar:static_cast<uint8_t>(m_data.networkServiceType)] forKey:@"networkServiceType"];
 
-    SET_DICT_FROM_PRIMITIVE(requestPriority, NSNumber, Int);
+    int clampedRequestPriority = std::min(std::max(m_data.requestPriority, -1), 4);
```
I think it would be nice to either call toResourceLoadPriority here or use ResourceLoadPriority enum values so we don't have mysterious magic numbers we may forget to update if we ever, for example, add another priority.

pedrovarangot replied:
But aren't this technically `CFURLRequestPriority` or -1? Or did I misunderstand that? `ResourceLoadPriority` is 0 to 4 and `toResourceLoadPriority` is the function that will assert if this value is out of range.

pedrovarangot replied:
The code that had an `enum CFURLRequestPriority` is commented out on the header and it's now a `CFIndex`, if I trust the code of `toResourceLoadPriority` I really can't find a way to not hardcode this values there. I don't understand the FIXME comment there or how the radar they reference affects the -1 to 4 or 0 to 5 mapping, plus even if that did, they are still defining `CFURLRequestPriority` as a CFIndex and I can't find where the values are defined to -1 to 4 except on our own function.

achristensen07 replied:
I was thinking something like this:
std::min(std::max(m_data.requestPriority, -1), WTF::enumToUnderlyingType(ResourceLoadPriority::VeryHigh))

achristensen07 commented on LayoutTests/ipc/cfnetwork-crashes-with-string-to-string-http-headers.html:
```diff
@@ -38,7 +38,7 @@
                                         timeWindowDuration: 15,
                                         networkServiceType: 0,
-                                        requestPriority: 89,
+                                        requestPriority: 0,
```
This change should not be made.  We still need the test to verify what happens when out of bounds values are sent.

pedrovarangot replied:
I could add another test? this was testing a different crash.

achristensen07 replied:
It's ok for a test to test multiple crashes.  I think we should leave this test as it is instead of adding another test.

Commit:
REGRESSION(286883@main): ASSERT NOT REACHED in WebCore::toResourceLoadPriority(CFURLRequestPriority)
rdar://140442183
https://bugs.webkit.org/show_bug.cgi?id=283596

Reviewed by Ryosuke Niwa.

This fixes the regression by setting a value for the priority that is within bounds, and prevents future asserts by clamping

* LayoutTests/ipc/cfnetwork-crashes-with-string-to-string-http-headers.html:
* LayoutTests/ipc/url-requestPriority-out-of-range-expected.txt: Added.
* LayoutTests/ipc/url-requestPriority-out-of-range.html: Copied from LayoutTests/ipc/cfnetwork-crashes-with-string-to-string-http-headers.html.
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
(WebKit::CoreIPCNSURLRequest::toID const):

Canonical link: https://commits.webkit.org/289501@main
